File: c#/diffusion.cs
using System;

// Csharp checked on 10/20/17

public class Diffusion
{
    static public void Main()
    {
        int msize;
        Console.WriteLine("Enter the size of the box: ");
        msize = Convert.ToInt32(Console.ReadLine());
        Console.WriteLine("The Msize is: " + msize);
        if(msize > 0 && msize <100)
        {
        //Varibles that are the same for setting up the cube with or without the partition
        int N = msize;                               //maxsize
        double[,,] A = new double[N,N,N];               //cube declaration of size NxNxN
        int i,j,k,l,m,n;
        int wall = -1;                                  //the value/concentration of a wall block
        double change;                                  //is the change of concentrations between blocks
        double D = 0.175;                               //diffusion coefficient 
        double room = 5;                                //room dimension is 5 meters
        double speed = 250.0;                           //speed of gas molecule based on 100 g/mol gas at RT
        double timestep = (room/speed)/N;               //basis for spatial stepsizes with respect to position in seconds
        double disblock = room/N;                       //is the distance between blocks
        double DTerm = D*timestep/(disblock*disblock);  //
        double time = 0.0;                              //tracks the simulated time
        double ratio = 0.0;                             //the ratio of the min concentration/the max concentration

        Console.WriteLine("Attempting to allocate " + (N*N*N) + " Doubles");
        
        //User input: enter 1 to run with a partition, enter 2 to run without
        bool validAnswer = false; //used for choosing if the user wants a partition
        int p;
        
        while (!validAnswer) 
        {
            Console.WriteLine("To run program with partition, enter 1, to run without a partition enter 2: ");
            validAnswer = int.TryParse(Console.ReadLine(), out p); 
            if(p==1 || p==2)
            {
                if(p==2)
                {   //***start non-partition code***
                    Console.WriteLine("Running program without a partition...");
                    //set 3d array to 0.0 except for first cube which is set to 1.0e21
                    for(i=0;i<N;i++)
                    {
                        for(j=0;j<N;j++)
                        {
                            for(k=0;k<N;k++)
                            {
                                if(i==0 && j==0 && k==0)
                                {
                                    A[0,0,0] = 1.0e21;
                                }
                                else
                                {
                                    A[i,j,k] = 0.0;
                                }
                            }
                        }
                    }
                    
                    //Goes through each block and compares the concentration (ratio) of every block next to it.
                    //The change (change of concentration) between the ratios is subtracted from the original and added to the new,
                    //in order to show the concentration of the gas changes as simulated time continues.
                    while(ratio < 0.99)
                    {
                        for(i=0;i<N;i++)
                        {
                            for(j=0;j<N;j++)
                            {
                                for(k=0;k<N;k++)
                                {
                                    if(k+1 < N)
                                    {
                                                change=(A[i,j,k]-A[i,j,k+1])*DTerm;
                                                A[i,j,k]=A[i,j,k]-change;
                                                A[i,j,k+1]=A[i,j,k+1]+change;
                                    }
                                    if(k-1 >= 0)
                                    {
                                                change=(A[i,j,k]-A[i,j,k-1])*DTerm;
                                                A[i,j,k]=A[i,j,k]-change;
                                                A[i,j,k-1]=A[i,j,k-1]+change;
                                    }
                                    if(j+1 < N)
                                    {
                                                change=(A[i,j,k]-A[i,j+1,k])*DTerm;
                                                A[i,j,k]=A[i,j,k]-change;
                                                A[i,j+1,k]=A[i,j+1,k]+change;
                                    }
                                    if(j-1 >= 0)
                                    {
                                                change=(A[i,j,k]-A[i,j-1,k])*DTerm;
                                                A[i,j,k]=A[i,j,k]-change;
                                                A[i,j-1,k]=A[i,j-1,k]+change;
                                    }
                                    if(i+1 < N)
                                    {
                                                change=(A[i,j,k]-A[i+1,j,k])*DTerm;
                                                A[i,j,k]=A[i,j,k]-change;
                                                A[i+1,j,k]=A[i+1,j,k]+change;
                                    }
                                    if(i-1 >= 0)
                                    {
                                                change=(A[i,j,k]-A[i-1,j,k])*DTerm;
                                                A[i,j,k]=A[i,j,k]-change;
                                                A[i-1,j,k]=A[i-1,j,k]+change;
                                    }
                                }
                            }
                        }
                        time = time + timestep; //increments simulated time
                        //Check for mass consistency: to make sure that we are accounting for every molecule of gas, and that none of the gas goes 
                        //outside the cube. Determines the new ratio after every step: divides the new min by the new max.
                        double sumval = 0.0;    //sum of the concentrations of every block in the cube
                        double max = A[0,0,0];  //block with the highest concentration
                        double min = A[0,0,0];  //block with the lowest concentration

                        for(i=0;i<N;i++)
                        {
                            for(j=0;j<N;j++)
                            {
                                for(k=0;k<N;k++)
                                {
                                    if(max < A[i,j,k])
                                    {
                                        max = A[i,j,k]; //new max if a bigger concentration is found within the cube
                                    }
                                    if(min > A[i,j,k])
                                    {
                                        min = A[i,j,k]; //new min if a smaller concentration is found within the cube
                                    }
                                    sumval += A[i,j,k]; //adds the sum of every block in the cube
                                }
                            }
                        }
                        ratio = min/max;    //calculates the updated ratio for every step
                        Console.WriteLine(time + " " + ratio + " " + sumval);
                    }
                    Console.WriteLine("Box equilibrated in " + time + " seconds of simulated time without a partition.");
                }   //***end non-partition code***
                else if(p==1)
                {   //***start partition code***
                    Console.WriteLine("Running program with a partition...");
                    //set 3d array to 0.0 and set up the walls=-1
                    for(i=0;i<N;i++)
                    {
                        for(j=0;j<N;j++)
                        {
                            for(k=0;k<N;k++)
                            {
                                if((j >= (N/2)-1) && (k == (N/2)-1))
                                {
                                    A[i,j,k] = wall;
                                }
                                else
                                {
                                    A[i,j,k] = 0.0;
                                }
                            }
                        }
                    }
                    A[0,0,0] = 1.0e21;  //initializes the first cube to 1.0e21
            
                    //Goes through each block and compares the concentration (ratio) of every block next to it as long as the current block is not a wall.
                    //Also, the if statements check that a specific block next to the current is not a wall ex. A[i,j,k+1]!=-1.
                    //The change (change of concentration) between the ratios is subtracted from the original and added to the new, in order
                    //to show the concentration of the gas changes as simulated time continues.                    
                    
                    while(ratio <= .99)
                    {
                        for(i=0;i<N;i++)
                        {
                            for(j=0;j<N;j++)
                            {
                                for(k=0;k<N;k++)
                                {
                                    if(A[i,j,k] != -1)    //checks to make sure that the current block is not a wall
                                    {
                                        if(k+1 < N && k-1 >= 0)
                                        {
                                            if(A[i,j,k+1] != A[i,j,k] && A[i,j,k+1] != -1)
                                            {
                                                change=(A[i,j,k]-A[i,j,k+1])*DTerm;
                                                A[i,j,k]=A[i,j,k]-change;
                                                A[i,j,k+1]=A[i,j,k+1]+change;
                                            }
                                            if(A[i,j,k-1] != A[i,j,k] && A[i,j,k-1] != -1)
                                            {
                                                change=(A[i,j,k]-A[i,j,k-1])*DTerm;
                                                A[i,j,k]=A[i,j,k]-change;
                                                A[i,j,k-1]=A[i,j,k-1]+change;
                                            }
                                        }
                                        
                                        if(j+1 < N && j-1 >= 0)
                                        {
                                            if(A[i,j+1,k] != A[i,j,k] && A[i,j+1,k] != -1)
                                            {
                                                change=(A[i,j,k]-A[i,j+1,k])*DTerm;
                                                A[i,j,k]=A[i,j,k]-change;
                                                A[i,j+1,k]=A[i,j+1,k]+change;
                                            }
                                            if(A[i,j-1,k] != A[i,j,k] && A[i,j-1,k] != -1)
                                            {
                                                change=(A[i,j,k]-A[i,j-1,k])*DTerm;
                                                A[i,j,k]=A[i,j,k]-change;
                                                A[i,j-1,k]=A[i,j-1,k]+change;
                                            }
                                        }   
                                        if(i+1 < N && i-1 >= 0)
                                        {
                                            if(A[i+1,j,k] != A[i,j,k] && A[i+1,j,k] != -1)
                                            {
                                                change=(A[i,j,k]-A[i+1,j,k])*DTerm;
                                                A[i,j,k]=A[i,j,k]-change;
                                                A[i+1,j,k]=A[i+1,j,k]+change;
                                            }
                                            if(A[i-1,j,k] != A[i,j,k] && A[i-1,j,k] != -1)
                                            {
                                                change=(A[i,j,k]-A[i-1,j,k])*DTerm;
                                                A[i,j,k]=A[i,j,k]-change;
                                                A[i-1,j,k]=A[i-1,j,k]+change;
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    
                        time = time + timestep; //increments the wall time
                        
                        //Check for mass consistency: to make sure that we are accounting for every molecule of gas, and that none of the gas goes
                        //outside the cube. Determines the new ratio after every step: divides the new min by the new max.
                        double sumval = 0.0;    //sum of the concentrations of every block in the cube
                        double max = A[0,0,0];  //block with the highest concentration
                        double min = A[0,0,0];  //block with the lowest concentration
                    
                        for(i=0;i<N;i++)
                        {
                            for(j=0;j<N;j++)
                            {
                                for(k=0;k<N;k++)
                                {
                                    if(A[i,j,k] != wall)    //checks to make sure the current block is not a wall
                                    {
                                        if(max < A[i,j,k])
                                        {
                                            max = A[i,j,k]; //new max if a bigger concentration is found within the cube
                                        }
                                        if(min > A[i,j,k])
                                        {
                                            min = A[i,j,k]; //new min if a smaller concentration is found within the cube
                                        }
                                        sumval += A[i,j,k]; //adds the sum of every block in the cube
                                    }
                                }
                            }
                        }
                
                        ratio = min/max;    //calculates the updated ratio for every step
                        Console.WriteLine(time + " " + ratio + " " + sumval);
                    }
               
                Console.WriteLine("Box equilibrated in " + time + " seconds of simulated time with a partition.");
                }
            }   //***ends partition code***
            else
            {
                validAnswer=false;
            }
        }
    }
    else
    {
        Console.WriteLine("Entered an invalid input");
    }
    }
}

File: go/diffusion.go
package main
// Go code checked on 10/20/17
import (
    "fmt"
    "strings"
    "time"
)
//function to run diffusion
func diffusion() {
    //Varibles that are the same for setting up the cube with or without the partition
    const N int = 10                        //maxsize
    var A[N][N][N]float64                   //cube declaration of size NxNxN
    var i,j,k,l,m,n int

    x := 10.0                               //N but a double
    D := 0.175                              //diffusion coefficient
    room := 5.0                             //room dimension is 5 meters
    speed := 250.0                          //speed of gas molecule based on 100 g/mol gas at RT
    timestep := (room/speed)/x              //basis for spatial stepsizes with respect to position in seconds
    disblock := room/x                      //is the distance between blocks
    DTerm := D*timestep/(disblock*disblock) //
    time := 0.0                             //tracks simulated time
    ratio := 0.0                            //the ratio of the min concentration/the max concentration

    //User input: enter yes in order to run the program with a partition. The default case will run without a partition.
    fmt.Print("To run program with a partition please enter 'yes', to run without a partition please enter 'no': ")
    var input string
    var lowerinput string
    fmt.Scanln(&input)
    lowerinput = strings.ToLower(input) //changes the input to lower case
    fmt.Println(strings.ToLower(input))
    
    switch lowerinput {
        case "yes":
            //***start of partition code***
            fmt.Println("This program will run with a partition...")

            //initializes the initial values of the cube as well as the walls
            for i=0; i < N; i++ {
                for j=0; j<N; j++ {
                    for k=0; k<N; k++ {
                        if i==0 && j==0 && k==0 {
                            A[0][0][0] = 1.0e21
                        } 
                        if j>=(N/2)-1 && k==(N/2)-1 {
                            A[i][j][k] = -1       //-1 is the value of a wall
                        } else {
                            A[i][j][k] = 0.0
                        }
                    }
                }
            }
            A[0][0][0] = 1.0e21     //initializes the first cube to 1.0e21
            
            //Goes through each block of two versions of the room (original[i][j][k]) and new([l][n][m])) and compares the concentration (ratio) 
            //of every block next to it as long as the current block is not a wall. Also, the if statement containing all of the cases has an extra 
            //case to check if that specific block next to the current is not a wall ex. A[l][m][n+1]!=-1. The change (change of concentration) 
            //between the ratios is subtracted from the original and added to the new, in order to show the concentration of the gas changes 
            //as simulated time continues.

            for ratio<0.99 {
                time = time + timestep  //increments the simulated time
                for i=0; i<N; i++ {
                    for j=0; j<N; j++ {
                        for k=0; k<N; k++ {
                            for l=0; l<N; l++ {
                                for m=0; m<N; m++ {
                                    for n=0; n<N; n++ {
                                        if A[l][m][n] != -1 {   //checks that the current block is not a wall
                                            if((i==l)&&(j==m)&&(k==n+1)&&(A[l][m][n+1]!=-1))||((i==l)&&(j==m)&&(k==n-1)&&(A[l][m][n-1]!=-1))||((i==l)&&(j==m+1)&&(k==n)&&(A[l][m+1][n]!=-1))||((i==l)&&(j==m-1)&&(k==n)&&(A[l][m-1][n]!=-1))||((i==l+1)&&(j==m)&&(k==n)&&(A[l+1][m][n]!=-1))||((i==l-1)&&(j==m)&&(k==n)&&(A[l-1][m][n]!=-1)) {
                                                change := (A[i][j][k]-A[l][m][n])*DTerm;
                                                A[i][j][k] = A[i][j][k] - change;
                                                A[l][m][n] = A[l][m][n] + change;
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                } 

                //Check for mass consistency: to make sure that we are accounting for every molecule of gas, and that none of the gas goes outside the cube.
                //Determines the new ratio after every step: divides the new min by the new max.
                sumval := 0.0         //sum of the concentrations of every block in the cube
                max := A[0][0][0]     //block with the highest concentration
                min := A[0][0][0]     //block with the lowest concentration

                for i=0; i<N; i++ {
                    for j=0; j<N; j++ {
                        for k=0; k<N; k++ {
                            if A[i][j][k] != -1 {
                                if max < A[i][j][k] {
                                    max = A[i][j][k]    //new max if a bigger concentration is found within the cube
                                }
                                if min > A[i][j][k] {
                                    min = A[i][j][k]    //new min if a smaller concentration is found within the cube
                                }
                                sumval += A[i][j][k]    //adds the sum of every block in the cube
                            }
                        }
                    }
                }
                ratio = min/max     //calculates the updated ratio for every step

                fmt.Printf("%g %g %g\n", time, ratio, sumval)
            }
            fmt.Printf("Box equilibrated in %g seconds of simulated time with a partition.\n", time)
        //***end of partition code***

        //***start of non-partition code***
        case "no":
            fmt.Println("This program will run without a partition...")
    
            //initializes the first cube to 1.0e21 and the others to 0.0
            for i=0; i < N; i++ {
                for j=0; j<N; j++ {
                    for k=0; k<N; k++ {
                        if i==0 && j==0 && k==0 {
                            A[0][0][0] = 1.0e21
                        } else {
                            A[i][j][k] = 0.0
                        }
                    }
                }
            }
    
            //Goes through each block of two versions of the room (original[i][j][k]) and new([l][n][m]))and compares the concentration (ratio)
            //of every block next to it. The change (change of concentration) between the ratios is subtracted from the original and added to the new,
            //in order to show the concentration of the gas changes as simulated time continues.

            for ratio<0.99 {
                time = time + timestep  //increments the simulated time
                for i=0; i<N; i++ {
                    for j=0; j<N; j++ {
                        for k=0; k<N; k++ {
                            for l=0; l<N; l++ {                    
                                for m=0; m<N; m++ {
                                    for n=0; n<N; n++ {
                                        if ((i==l)&&(j==m)&&(k==n+1))||((i==l)&&(j==m)&&(k==n-1))||((i==l)&&(j==m+1)&&(k==n))||((i==l)&&(j==m-1)&&(k==n))||((i==l+1)&&(j==m)&&(k==n))||((i==l-1)&&(j==m)&&(k==n)) {
                                            change := (A[i][j][k]-A[l][m][n])*DTerm
                                            A[i][j][k] = A[i][j][k] - change
                                            A[l][m][n] = A[l][m][n] + change
                                        }
                                    }
                                }
                            }
                        }
                    }
                }

                //Check for mass consistency: to make sure that we are accounting for every molecule of gas, and that none of the gas goes outside the cube.
                //Determines the new ratio after every step: divides the new min by the new max.
                sumval := 0.0         //sum of the concentrations of every block in the cube
                max := A[0][0][0]     //block with the highest concentration
                min := A[0][0][0]     //block with the lowest concentration

                for i=0; i<N; i++ {
                    for j=0; j<N; j++ {
                        for k=0; k<N; k++ {
                            if max < A[i][j][k] {
                                max = A[i][j][k]    //new max if a bigger concentration is found within the cube
                            }
                            if min > A[i][j][k] {
                                min = A[i][j][k]    //new min if a smaller concentration is found within the cube
                            }
                            sumval += A[i][j][k]    //adds the sum of every block in the cube
                        }
                    }
                }
                ratio = min/max     //calculates the updated ratio for every step

                fmt.Printf("%g %g %g\n", time, ratio, sumval)
            } 
            fmt.Printf("Box equilibrated in %g seconds of simulated time without a partition.\n", time)
            //***end of non-partition code***

        //if invalid input is entered
        default:
            fmt.Printf("Entered an invalid input")
    }    
}
//function to track the wall time of the diffusion function
func main() {
    start := time.Now()
    diffusion()
    end := time.Now()
    fmt.Printf("Wall Time:  %v\n", end.Sub(start))
  }
